### PR TITLE
Ignore images from embedded html when pasting, paste plaintext instead.

### PR DESCRIPTION
**Pull request by qispark**
### Details
When pasting some HTML containing images into Composer, it will cause an error modal. This is because of the server's Content Security Policy, which prevents those images from being fetched separately. This PR prevents the images from being fetched, instead it will paste the plaintext that is copied.

### Fixed Issues
$ https://github.com/Expensify/App/issues/18583
$ https://github.com/Expensify/App/issues/18583#issuecomment-1598515619

### Tests
1. Copy a web document which contains both text and images. (e.g. https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_image_test)
2. Log in with any account.
3. Go to any chat
4. Paste the contents into the composer
5. Verify that no image attachments are being created.
6. Verify that the correct contents are within the composer box.
7. Verify that no error modals are appearing upon paste.

- [x] Verify that no errors appear in the JS console

### Offline tests
N/A

### QA Steps
1. Copy a web document which contains both text and image. (e.g. https://www.w3schools.com/tags/tryit.asp?filename=tryhtml_image_test)
2. Log in
3. Go to any chat
4. Paste the contents into the composer
5. Verify that the correct contents are within the composer box.
6. Verify that no error modals are appearing upon paste.

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/31749903/21079b2a-ecc6-4f14-bc56-5e5f453725d6

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/31749903/44bdf909-4ef6-440d-965a-dfecb456a30d

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/31749903/a31a1cfd-454b-4495-a2bb-7ac4c96509b8

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/31749903/7af5aaaf-ab1f-4503-bfc2-b775ae854ce7

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/31749903/0e904d2b-df71-48c1-9c4a-c953baf0e37d

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/31749903/a3f7e7b9-251c-4b43-9ca0-86fa4b6c0b89

</details>

### CODE REVIEW

File: src/components/Composer/index.js
Comment (qispark) [reply src/components/Composer/index.js:324]: @parasharrajat If you read my proposal carefully, I explain the caveat of why we should paste plaintext instead of pasting the HTML (https://github.com/Expensify/App/issues/18583#issuecomment-1598515619)

If any images have links attached, and we try to paste them, it leads to formatting issues in the markdown that is generated.

Please let me know what is the desired behaviour in this instance as I thought you already approved my proposal.

Comment (qispark) [reply src/components/Composer/index.js:324]: This is what happens when I set it to paste HTML:

https://github.com/Expensify/App/assets/31749903/7c9313d6-36b9-4348-8e44-81b12a186c0d

Comment (qispark) [reply src/components/Composer/index.js:324]: @parasharrajat This is only when the images are wrapped in a link.

Comment (techievivek) [reply src/components/Composer/index.js:324]: > It is strange that image tags are converted to links
It would be because the image are used as links? @qispark Can you test the same on Slack and WhatsApp and see what is being pasted there? But I agree with @parasharrajat that if the copied content has HTML then we would want to paste them as HTML (MINUS the image part).

Comment (qispark) [reply src/components/Composer/index.js:324]: @techievivek The images are stripped on platforms like Slack.

The only way forward then is to update ExpensiMark, as it is formatting the pasted content into Markdown. I also mentioned this in my original comment (https://github.com/Expensify/App/issues/18583#issuecomment-1598515619).

I think if we can implement a new function in ExpensiMark, which does the same thing (e.g. `parser.htmlToMarkdownWithoutImages`) we can utilise it here. What are your thoughts, and is this a feasible option?

Comment (qispark) [reply src/components/Composer/index.js:324]: @parasharrajat Let me know if I should action on anything based on your feedback.

Comment (techievivek) [reply src/components/Composer/index.js:324]: @qispark Based on Rajat's comment,` If images are wrapped in a link then it is fine if they are parsed as links in the pasted code.` I think the image was wrapped in a link in the recording above, right? (Can you confirm this by inspecting that element?) and when we are pasting it is being rendered as links.

@parasharrajat
> But if the image itself is being converted to links then that is a problem

I don't _think_ that is happening in this case. @qispark can you please confirm this?


And now for the last part

> Why are we not pasting the HTML? embeddedImages[0].dataset && embeddedImages[0].dataset.stringifyType === 'emoji' require pasting text but other than this we can use html pasting.

@qispark What do you think can be done to fix this? Do we need to update the ExpensiMark logic? I think that is very much doable, so if that's the case please push a PR in the expensify-common repo and let's keep this moving.

**Comment by qispark** [comment]
OK, @parasharrajat I'm happy to wait until that PR is merged before we test this again - we might not need further changes in that case. Is that ok?

**Comment by qispark** [comment]
@parasharrajat I have just updated the PR to reinstate the logic for emojis. Please let me know if any other changes are required.

**Comment by techievivek** [comment]
https://github.com/Expensify/expensify-common/pull/555 PR is merged, so I think we can move forward with thoroughly testing this PR now, right?

**Comment by qispark** [comment]
@techievivek I've just tested it. Unfortunately, it appears to have the same behaviour, still creating an empty link. I'm using a test branch based on `main` with this PR's updates.

Do we need to revise ExpensiMark again? @parasharrajat 

Screenshot 2023-07-07 at 3 32 32 pm

**Comment by qispark** [comment]
In `lib/ExpensiMark.js`:
```
{
                name: 'link',
                process: (textToProcess, replacement) => this.modifyTextForUrlLinks(MARKDOWN_LINK_REGEX, textToProcess, replacement),
                replacement: (match, g1, g2) => {
                    if (g1.match(CONST.REG_EXP.EMOJIS) || !g1.trim()) {
                        return match;
                    }
                    return `<a href="${Str.sanitizeURL(g2)}" target="_blank" rel="noreferrer noopener">${g1.trim()}</a>`;
                },
            },
```

If this is the relevant rule, I think we need to update the part `!g1.trim()` to strip new lines as well in addition to just whitespaces. It's the new line causing the empty link to appear, I believe.

**Comment by qispark** [comment]
@parasharrajat I'm OOO until Monday - but essentially you can recreate it on any website with an image wrapped in a link.

**Comment by qispark** [comment]
@parasharrajat 
Here's an example I found, which creates an empty link, even without a new line:

```
<meta charset='utf-8'><p style="color: rgb(0, 0, 0); font-family: Times; font-size: medium; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal; text-decoration-thickness: initial; text-decoration-style: initial; text-decoration-color: initial;">The image below is a link. Try to click on it.</p><a href="https://www.w3schools.com/html/default.asp" style="font-family: Times; font-size: medium; font-style: normal; font-variant-ligatures: normal; font-variant-caps: normal; font-weight: 400; letter-spacing: normal; orphans: 2; text-align: start; text-indent: 0px; text-transform: none; widows: 2; word-spacing: 0px; -webkit-text-stroke-width: 0px; white-space: normal;"><img src="https://www.w3schools.com/html/smiley.gif" alt="HTML tutorial" style="width: 42px; height: 42px;"></a>
```

You can copy the contents from here: https://www.w3schools.com/html/tryit.asp?filename=tryhtml_links_image

This is the result:

Screenshot 2023-07-10 at 3 35 23 pm

**Comment by qispark** [comment]
Hi @parasharrajat, it wasn't clear to me, but did you want me to make a PR in expensify-common? Or shall I wait until you have tested it?

**Comment by techievivek** [comment]
@parasharrajat, any updates here?

**Comment by techievivek** [comment]
@parasharrajat Gentle bump to keep this moving, also QA reported that this was not reproducible https://github.com/Expensify/App/issues/18583#issuecomment-1649115793 🤯, haven't verified it myself tho.

**Comment by techievivek** [comment]
@parasharrajat, Can you please prioritize this PR, as it has been open for some time now? Thanks

**Comment by qispark** [comment]
@parasharrajat From our last conversation, we were discussing some proposed changes to ExpensiMark. We already established that this PR alone also isn't fixing the issue. I think you were going to review something before I tried to make a PR there.

**Comment by qispark** [comment]
@parasharrajat Main is merged in now. Let me know if you need anything else.

**Review by parasharrajat — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.


:ribbon: :eyes: :ribbon:  C+ reviewed

**Comment by techievivek** [comment]
Sorry, I was ooo on Friday, taking a look now. Thanks.

File: src/components/Composer/index.js
Comment (techievivek) [inline src/components/Composer/index.js:350]: @qispark Looks like we have updated our CSP rule now to all fetching images from everywhere `img-src * blob: data:;`. Do you think this would now allow us to load the images from the pasted content correctly?

**Comment by techievivek** [comment]
@qispark Linting seems to be failing, can you please fix that, thanks.

File: src/components/Composer/index.js
Comment (techievivek) [reply src/components/Composer/index.js:350]: I tested copying pasting content with images but even with CSP changes this doesn't work. Can you look into it so we update the comments correctly?

**Comment by qispark** [comment]
@techievivek Just updated lint.

File: src/components/Composer/index.js
Comment (qispark) [reply src/components/Composer/index.js:339]: @techievivek I was on Staging and still found a CSP issue as it currently stands. Can you let me know which environment has updated CSP and I can try and replicate?

Comment (techievivek) [reply src/components/Composer/index.js:339]: We have updated for both the platforms, you can confirm this by looking at the headers in response.

Comment (qispark) [reply src/components/Composer/index.js:339]: OK. I have done some testing. The reason the CSP error is still occurring is related to `connect-src` rather than `img-src`. As we are trying to use the Fetch API, rather than fetching within an image tag, it relates to the `connect-src` policy.

Comment (techievivek) [reply src/components/Composer/index.js:339]: Aah, nice, that makas sense now. Thanks

Comment (techievivek) [reply src/components/Composer/index.js:339]: Can you please update the content to specify content-src CSP so it is easy to follow that in future.

File: src/components/Composer/index.js
Comment (techievivek) [inline src/components/Composer/index.js:339]: Maybe?
```suggestion
            // Exclude parsing img tags in the HTML, as fetching the image via fetch triggers a connect-src Content-Security-Policy error.
```

Comment (qispark) [reply src/components/Composer/index.js:339]: LGTM

**Review by techievivek — DISMISSED** [review]
Great job, thanks for sticking by 🙇

**Comment by qispark** [comment]
@techievivek Just updated again due to prettier failing.

**Review by techievivek — APPROVED** [review]

**Pull request merged**
